### ISSUE TITLE
Java 21 Masterplan

### ISSUE THREAD

**Issue by cdietrich** [eclipse/xtext#2686]
Once Java 21 LTS is out we should try to support it

- [ ] set minimum supported version to Java 17, Drop Java 11 (build)
- [ ] adapt JavaVersion Enum
- [ ] adapt Wizard tests
- [ ] adapt platform dependencies in dev-bom
- [ ] adapt minimal target platform
- [ ] adapt gradle version in wizard
- [ ] check if classgraph needs adaptation?
- [ ] check if tycho needs adaptation?
- [ ] Guice/Guava update needed?
- [ ] new MWE?
- [ ] check java version switches in tests
- [ ] new Jetty version needed?
- [ ] with new minimal target version, check which reflective switches we can get rid of (platform, jdt)
- [ ] deal with Java deprecations

TODO: 100 things i have forgot

**Comment by cdietrich** [comment]
Guice and Guava is now on latest. we will see if this is sufficient

**Comment by OLibutzki** [comment]
As Java 21 has been released today, do you have any roadmap when to support it? Especially solving the jdt issue mentioned in #2680 is something we are looking forward to.

**Comment by cdietrich** [comment]
There is no road map and as Xtext is no longer my day job it might take long. First question: are Java 21 capable eclipse platform and jdt maven artifacts already available or will they ship in December ?

**Comment by OLibutzki** [comment]
Hmm, this issue us "open", no new comments for months.

https://github.com/eclipse-jdt/eclipse.jdt.core/issues/890

**Comment by cdietrich** [comment]
https://github.com/eclipse-jdt/eclipse.jdt.core/commits/BETA_JAVA21 Shows some wip

**Comment by mpalat** [comment]
Please see : https://www.eclipse.org/lists/eclipse-dev/msg12203.html

**Comment by cdietrich** [comment]
with https://www.eclipse.org/lists/eclipse-dev/msg12207.html we wont have maven artifacts before end of year.
am not sure if there are maven snapshots meanwhile so development can start before

**Comment by DanySK** [comment]
Hi, is there any update on this plan?

I guess that errors like `org.eclipse.xtext.common.types.access.TypeResource$CannotLoadTypeResourceException: java.lang.IllegalArgumentException: Unsupported class file major version 65` are due to Xtext not yet supporting Java 21 - am I correct?

I am asking because this is preventing [Protelis](https://protelis.github.io) from becoming Java 21 compliant which, in turn, also prevents Java 21 compliance for the [Alchemist Simulator](https://alchemistsimulator.github.io).

**Comment by cdietrich** [comment]
No jdt java 21 maven artifacts no java 21 in Xtext

**Comment by cdietrich** [comment]
Besides that: when these become available this is a help wanted story

**Comment by gmlongo1** [comment]
Has there been any progress on this? We make excessive use of generators in our daily builds and are looking to move to Java 21 in the very near future.

**Comment by cdietrich** [comment]
Obviously no. And Xtext is no longer my day job. And nobody seems to step in

**Comment by gmlongo1** [comment]
Not sure why "obviously" was needed, as its been 4+ months so we were just looking for any feedback. Are we saying xtext is no longer being officially maintained? If there is no roadmap for Java 21+, those of us that heavily rely on xtext will have to fmake plans to move away from the tool.

**Comment by cdietrich** [comment]
https://github.com/eclipse/xtext/issues/1721

I don’t think I alone can do this in my spare time while working at a startup

**Comment by gmlongo1** [comment]
I completely understand the lack of time due to other commitments. We will need to react accordingly.

**Comment by LorenzoBettini** [comment]
This doesn't mean that Xtext is no longer officially maintained (see the number of commits in the last few months).

However, IIRC, if you don't use records from within xtend you shouldn't have problems switching to Java 21, or am I missing something?

**Comment by gmlongo1** [comment]
When I switch to Java 21 from 17, I start getting "cannot resolve proxy: java:/objects/java.lang.string#java.lang.string" errors.

**Comment by cdietrich** [comment]
The  old jdt dependencies and the Java version Enum don’t support Java 21

**Comment by DanySK** [comment]
@LorenzoBettini nope, Xtext is failing with Java 21. It does not build with Java 21 (see: https://github.com/Protelis/Protelis-Parser/pull/345), and code produced with previous Java versions fails at runtime when running Java 21, see: https://scans.gradle.com/s/l4r2lg34vs5cu

`DiagnosticWrappedException: org.eclipse.xtext.common.types.access.TypeResource$CannotLoadTypeResourceException: java.lang.IllegalArgumentException: Unsupported class file major version 65`

**Comment by HannesWell** [comment]
> DiagnosticWrappedException: org.eclipse.xtext.common.types.access.TypeResource$CannotLoadTypeResourceException: java.lang.IllegalArgumentException: Unsupported class file major version 65

Can you provide a full Stack-Trace for that?
Maybe ASM needs an update.

**Comment by LorenzoBettini** [comment]
> The old jdt dependencies and the Java version Enum don’t support Java 21 So it wound work with source and target 21
> 

The old jdt dependencies are the maven ones, right? Because in the TP we already use the latest ones, don't we?

**Comment by cdietrich** [comment]
Yes the maven ones

**Comment by DanySK** [comment]
@HannesWell 
I did, can you access the gradle build scan? https://scans.gradle.com/s/l4r2lg34vs5cu

Here's how to reproduce:
1. `git clone https://github.com/Protelis/Protelis.git`
2. `cd Protelis`
3. `git checkout renovate/org.danilopianini.multi-jvm-test-plugin-0.5.x`
4. `./gradlew testWithJvm21`

It should "fail correctly" with any major OS and regardless of the JVM you use to run Gradle.

**Comment by HannesWell** [comment]
> I did, can you access the gradle build scan? https://scans.gradle.com/s/l4r2lg34vs5cu

Sorry I oversaw that, but looking at it now I fail to see the root cause.

I haven't checked-out your project, but I run all our xbase-based DSL with Java-21 as well as my developer eclipse (also executed mwe workflows) and didn't notice any issues with it.

The only thing I noticed so far(there might be more not yet uncovered) is that the `xtend-Maven-plugin` has a literal for Java-21 missing in the `JavaVersions` class.
I can provide a PR to add that.

**Comment by cdietrich** [comment]
as elaborated before

- enum needs adaptation
- jdt and platform dependencies need bump
- this also requires java 11 drop

**Comment by DanySK** [comment]
@HannesWell I fixed it by constraining the ASM version manually like this: https://github.com/Protelis/Protelis/commit/c78854fbe39058d1317b62b145be6c9a5996477f

**Comment by LorenzoBettini** [comment]
@HannesWell I might need your help: I started to experiment with Java 21 in my fork with GitHub Actions. I set Java 21 in GitHub Actions and it is indeed used

```
Apache Maven 3.9.6 (bc0240f3c744dd6b6ec2920b3cd08dcc295161ae)
Maven home: /home/runner/.m2/wrapper/dists/apache-maven-3.9.6-bin/3311e1d4/apache-maven-3.9.6
Java version: 21.0.2, vendor: Eclipse Adoptium, runtime: /usr/lib/jvm/temurin-21-jdk-amd64
```

However, Tycho prints this:

```
[WARNING] Using JavaSE-17 to fulfill requested profile of JavaSE-11. This might lead to faulty dependency resolution, consider defining a suitable JDK in the toolchains.xml.
[INFO] Resolving target definition file:/home/runner/work/xtext/xtext/org.eclipse.xtext.xbase.lib/../xtext-latest.target for environments=[macosx/cocoa/x86_64, win32/win32/x86_64, linux/gtk/x86_64]
```

Java 21 is mentioned when resolving the TP, but it says that it's using JavaSE-17.

I see this https://tycho.eclipseprojects.io/doc/4.0.6/tycho-compiler-plugin/compile-mojo.html#useJDK:

> Which JDK to use for compilation. Default value is SYSTEM which means the currently running JDK. If BREE is specified, MANIFEST header Bundle-RequiredExecutionEnvironment is used to define the JDK to compile against.

Since we don't configure the Tycho compiler plugin by default (and by default the maven toolchain plugin is not used), why is it using Java 17 and not Java 21 which is the system one? Have you experienced this in the past?

**Comment by cdietrich** [comment]
I assume it is related to toolchains xml

**Comment by LorenzoBettini** [comment]
@cdietrich I've updated also the toolchains.xml, but the reference to that file must be passed explicitly as we do in Jenkinsfile when forcing the use of Java 11, otherwise that doesn't come into play. In GitHub Actions that is not used.

**Comment by LorenzoBettini** [comment]
Just a few more findings: when using Java 21 the build fails for this reason, as far as I understand:

- Xtend seems to detect that the runtime is Java 21 and it generates this Java code:
```java
if ((equalIndentationChange && (!Objects.equal(leafs.getLeafs().getLast(), leaf)))) {
```
instead of this
```java
if ((equalIndentationChange && (!Objects.equal(IterableExtensions.<LeafInfo>last(leafs.getLeafs()), leaf)))) {
```
Because it detects the new added method in Java 21 `getLast`. At least, Xtend seems to detect the new Java API ;)
- the compilation of the Java code fails because it looks like Java 17 is being used, so the new method is not found

I opened https://github.com/eclipse-tycho/tycho/issues/3731 to verify whether I misunderstand how the Tycho compiler uses the current JDK.

**Comment by cdietrich** [comment]
Yes this is what I saw here 
https://github.com/xtext/xtext-reference-projects/tree/cd_someJ21tests

**Comment by HannesWell** [comment]
> Since we don't configure the Tycho compiler plugin by default (and by default the maven toolchain plugin is not used), why is it using Java 17 and not Java 21 which is the system one? Have you experienced this in the past?

From the top of my head I cannot say for sure why this happens.
Can you point me to your branch and the build logs of your run? On the master I see only Java-17 setup in the GH actions so I assume you didn't just replaced 17 by 21 if you have both?
When using GH workflows you have to pay attention because the setup-java action by default also generates a toolchains.xml nevertheless by default the tycho-compiler-plugin should not use that (and you actually have to ensure the jdk-tool identifiers follow the `JavaSE-<version>` naming schema).

**Comment by LorenzoBettini** [comment]
@HannesWell this is the branch I'm working on https://github.com/LorenzoBettini/xtext/tree/lb_java_21

note also the issue I created for Tycho above.

By the way, the toolchains created by the setup java action looks like that:

```xml
Run cat ~/.m2/toolchains.xml
<?xml version="1.0"?>
<toolchains xmlns="http://maven.apache.org/TOOLCHAINS/1.1.0"
  xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
  xsi:schemaLocation="http://maven.apache.org/TOOLCHAINS/1.1.0 https://maven.apache.org/xsd/toolchains-1.1.0.xsd">
  <toolchain>
    <type>jdk</type>
    <provides>
      <version>21</version>
      <vendor>temurin</vendor>
      <id>temurin_21</id>
    </provides>
    <configuration>
      <jdkHome>/opt/hostedtoolcache/Java_Temurin-Hotspot_jdk/21.0.2-13/x64</jdkHome>
    </configuration>
  </toolchain>
</toolchains>
```

However, although I created this new profile, which is correctly activated

```xml
<profile>
	<id>strict-jdk-21</id>
	<activation>
		<jdk>[21,)</jdk>
	</activation>
	<build>
		<plugins>
			<plugin>
				<groupId>org.apache.maven.plugins</groupId>
				<artifactId>maven-toolchains-plugin</artifactId>
				<version>3.1.0</version>
				<executions>
					<execution>
						<goals>
							<goal>toolchain</goal>
						</goals>
					</execution>
				</executions>
				<configuration>
					<toolchains>
						<jdk>
							<version>21</version>
						</jdk>
					</toolchains>
				</configuration>
			</plugin>
		</plugins>
	</build>
</profile>
```

The build now complains like that

```
[INFO] --- toolchains:3.1.0:toolchain (default) @ xtext-parent ---
[INFO] Required toolchain: jdk [ version='21' ]
Error:  No toolchain matched from 1 found for type jdk
Error:  Cannot find matching toolchain definitions for the following toolchain types:
jdk [ version='21' ]
```

is the toolchain plugin so strict that it also wants the exact matching vendor?

**Comment by LorenzoBettini** [comment]
Further updates: I managed to force Java 21 in the toolchains (the one of the `setup-java` action doesn't seem to work, https://github.com/actions/setup-java/issues/617).

However, the error issued by tycho-compiler is still the same. On a second thought, that's expected and I guess it's independent from my previous problem about the wrong Java version being used.

The error is produced correctly: since we use the properties

```xml
<maven.compiler.target>11</maven.compiler.target>
<maven.compiler.release>11</maven.compiler.release>
```

we are telling the Java compiler not to use new APIs introduced in Java > 11 and it does not find the new method `getLast`, which was introduced in Java 21.

Basically Xtend generates Java code taking into consideration the new Java 21 API and produces Java code that uses the new API. Then, our Java compilation settings makes the Java compiler fail when it tries to compile the generated Java code.

@cdietrich @szarnekow either we make Xtend consider the `maven.compiler.release` avoiding using newer Java API (which I guess it's hard) or we increase the `maven.compiler.release` or remove it completely. Since we use the toolchain in Jenkins already, removing the "release" option seems safe.

**Comment by HannesWell** [comment]
> Further updates: I managed to force Java 21 in the toolchains (the one of the setup-java action doesn't seem to work, actions/setup-java#617).

Sorry I didn't have more to help you with this this evening and couldn't verify your problem, but as a quick note: for tycho IIRC it is usually important to set up the identifier specially as for example in:

https://github.com/eclipse-equinox/equinox/blob/b327cae6d2b4181fb426a91f9fb7196029a88b71/.github/workflows/build.yml#L25-L35

**Comment by LorenzoBettini** [comment]
@HannesWell no problem and thanks anyway. However, in your example, you don't use the maven toolchains plugin, which seems not to like the `toolchains.xml` file generated by the setup action. It works if I refer to our file, `--toolchains releng/toolchains.xml`, which I have updated for Java 21, so that problem is fixed.

@szarnekow @cdietrich the problem with the Java compiler is fixed as well: it's enough to redefine `maven.compiler.release` to 21 only in the profile for Java 21:

```xml
<profile>
	<id>strict-jdk-21</id>
	<activation>
		<jdk>[21,)</jdk>
	</activation>
	<properties>
		<!--
			Since the Xtend compiler generates Java code using new API introduced in Java 21,
			so we must increase the release level or the Java compilation would not find the
			new API when it compiles the Java code generated by Xtend.
			see https://github.com/eclipse/xtext/issues/2686#issuecomment-2038175179
		-->
		<maven.compiler.release>21</maven.compiler.release>
	</properties>
	<build>
		<plugins>
			<plugin>
				<groupId>org.apache.maven.plugins</groupId>
				<artifactId>maven-toolchains-plugin</artifactId>
				<executions>
					<execution>
						<goals>
							<goal>toolchain</goal>
						</goals>
					</execution>
				</executions>
				<configuration>
					<toolchains>
						<jdk>
							<version>21</version>
						</jdk>
					</toolchains>
				</configuration>
			</plugin>
		</plugins>
	</build>
</profile>
```

Now the Java code generated by Xtend compiles fine.

And now we get to test failures: looks like the new Java 21 API for https://docs.oracle.com/en/java/javase/21/docs/api/java.base/java/util/List.html#getLast() is tormenting us ;)

At least the Xbase type system tests that were relying on `getLast` NOT being present in Java are now failing. Those tests were using `last` as extension method, but now `getLast` is found in the Java API and so those tests fail.
For example https://github.com/search?q=repo%3Aeclipse%2Fxtext%20testBounds_19&type=code

Example of failure:

```
org.eclipse.xtext.xbase.tests.typesystem.RecomputingBatchTypeArgumentTest.testBounds_19 -- Time elapsed: 0.010 s <<< FAILURE!
org.junit.ComparisonFailure: Iterable<? super Integer> vs ? super Integer expected:<[Iterable<? super Integer>]> but was:<[? super Integer]>
```

Also some content assist tests fail because they relied on the methods of List, which now are more numerous.
For example, https://github.com/eclipse/xtext/blob/4f1669f20b5f2bdd0ad5d4663b2b9261753fd389/org.eclipse.xtext.xbase.web/src/test/java/org/eclipse/xtext/xbase/web/test/ContentAssistTest.java#L125

Of course, I could update all the tests, which will then work only on Java 21.

I was wondering if we could refactor those tests NOT to rely on Java API but on Java code that we store in our tests. That way, those tests would not be fragile and would not break when the Java API evolves.

**Comment by LorenzoBettini** [comment]
@szarnekow OK, I'll try to work on that. I'm more concerned about those about Xbase type system: are those strictly relying on extension methods of our iterable extensions or could be done without? I mean, are they meant to test typing in the presence of extension methods? If you could please provide some hints in that respect.

**Comment by shankar-itemis** [comment]
Hi @cdietrich I see the check list for migration to java 21. I believe it also includes good to have features. what would be the mandatory features / fixes to be made in xtext to support java 21?
Background: I am trying to test xtext for java 21 and currently I see the xtext project and code generated are backward compatible. I got warning for annotated classes which were using 'javax.inject' or 'javax.annotation' package. But I guess thats from the Eclipse E4 Platform which will remove support for those javax-annotations in future.

**Comment by cdietrich** [comment]
@shankar-itemis which places make use of it ?

**Comment by shankar-itemis** [comment]
@cdietrich Around 1818 classes in xtext makes use of annotations. The classe names are attached as report. 

disclaimer: not all the classes were verified manually this report was generated by a utility app developed.
[annotatedClasses.pdf](https://github.com/eclipse/xtext/files/14885931/annotatedClasses.pdf)

**Comment by LorenzoBettini** [comment]
> I got warning for annotated classes which were using 'javax.inject' or 'javax.annotation' package. But I guess thats from the Eclipse E4 Platform which will remove support for those javax-annotations in future.

@shankar-itemis I think we removed `javax.inject` from Xtext codebase (https://github.com/search?q=repo%3Aeclipse%2Fxtext%20javax.inject&type=code). I always get that warning as well, but it must come from other bundles coming from other dependencies, I don't think it comes from Xtext.

**Comment by HannesWell** [comment]
> I always get that warning as well, but it must come from other bundles coming from other dependencies, I don't think it comes from Xtext.

Yes that warning is issued if javax-annotations are present in the application.
In general it could be made more specific but I didn't have the time for that yet (as it is also just a warning). And you can suppress that warning, as it is mentioned in the message.

**Comment by shankar-itemis** [comment]
> Can you provide the util along @shankar-itemis i dont really get / understand a lot of the matches in your list

Pls find the sources here https://github.com/shankar-itemis/annotationFinderUtil

**Comment by shankar-itemis** [comment]
> > I got warning for annotated classes which were using 'javax.inject' or 'javax.annotation' package. But I guess thats from the Eclipse E4 Platform which will remove support for those javax-annotations in future.
> 
> @shankar-itemis I think we removed `javax.inject` from Xtext codebase (https://github.com/search?q=repo%3Aeclipse%2Fxtext%20javax.inject&type=code). I always get that warning as well, but it must come from other bundles coming from other dependencies, I don't think it comes from Xtext.

Yes thats right. I mainly looked only the annotations and its not specific to inject.

**Comment by cdietrich** [comment]
@shankar-itemis still don’t get your point. What do you try to find ?

**Comment by cdietrich** [comment]
@shankar-itemis  but why is having an annotation in general and `com.google.inject.Inject` bad. of course we have a billion of these

**Comment by shankar-itemis** [comment]
@cdietrich Its not bad. As you know the javax.inject has to be migrated to jakarta.inject in java 21. So as part of JDK 21 analysis this was done. 

Although guice abstracts this dependency and implement JSR-330 completely, the Provider of guice extends JSR 330. https://github.com/google/guice/wiki/JSR330. And from guice 7.0 it only supports jakarta.inject and not java.inject. So I was just ensuring that.

**Comment by cdietrich** [comment]
i still dont dont understand.
to my knowledge we dont use javax.inject anymore

**Comment by shankar-itemis** [comment]
@cdietrich  To my understaing the guice provider was extending javax.inject. This was an implicit dependency that guice had with javax.inject and I was in search of that. 

Pls correct me if I my understanding is wrong here.

**Comment by cdietrich** [comment]
> And from guice 7.0 it only supports jakarta.inject and not java.inject. So I was just ensuring that.

but this would mean they change their code
so i dont understnad why the guice inject class should be bad

**Comment by szarnekow** [comment]
It isn't bad and neither is a use of an annotation bad.
I think the ongoing analysis is based on bogus assumptions.

The code-base can be grep'ed easily for javax.inject or com.google.inject to find usages.

For Xtext, we decided to use exclusively com.google.inject types to configure the dependency injection container. That won't change.

**Comment by shankar-itemis** [comment]
I would like to clarify that neither the annotations or guice inject are bad. I was just analysing the places where annotations were made use and was there some direct references to javax.inject.

sorry if i have confused you on this. @cdietrich

**Comment by szarnekow** [comment]
@shankar-itemis thank's for the analysis and for sharing the code.

The gained information is unfortunately not providing any actionable insights, since annotations are commonly used in Xtext and processing all of them in an untargeted fashion is not helpful.

**Comment by LorenzoBettini** [comment]
As noted here https://github.com/eclipse/xtext/issues/2823#issuecomment-2072223679 at least from Eclipse with JDT access to Java 21 records seems to work out of the box from Xtend. I tested it manually. @szarnekow any hint on how to place a test in some of the existing test projects?

Then, I guess we need to adapt the standalone part, including the upgrade of JDT in the BOM. To test that, would it be enough to use the tests in our maven plugin projects? Or are there better places to test that?

Such experiments are in my fork https://github.com/LorenzoBettini/xtext/tree/lb_java_21-2

While https://github.com/eclipse/xtext/pull/2986 could already be merged

**Comment by szarnekow** [comment]
> @szarnekow any hint on how to place a test in some of the existing test projects?

I'd guard the test with a check that we have a Java21 available.
If so, use a record type from the JDK from Xtend source or put a Java record source file into an Xtend test projects side by side.

**Comment by LorenzoBettini** [comment]
@szarnekow OK, but does something like this https://github.com/LorenzoBettini/xtext/commit/3059e1021d7239e28c97859db291642619afef18 make sense, or is it useless?

**Comment by szarnekow** [comment]
I think it does make sense, but we should also have a test that uses a Java record from Xbase (can be an Xtend integration test).

**Comment by LorenzoBettini** [comment]
@szarnekow yes, I just wanted to start from the bottom and then up ;)

**Comment by LorenzoBettini** [comment]
Some nice updates:

- added a compilation test for Xbase accessing Java record https://github.com/LorenzoBettini/xtext/commit/c2befaf599a7cf52841f9118d2bc095640d3e0ef @szarnekow I haven't found a record in the JDK, unless I miss something, there's only a record about some Unix stuff, so I used draw2d that's already in our TP; I could add one also in Xtend but having in Xbase should be enough, shoudln't it? In Xtend, I'm planning to a add a validation test trying to extend a record, which is detected as final; by debugging, I found that a record is automatically considered final
- then a test for standalone builder https://github.com/LorenzoBettini/xtext/commit/ffac152fbb322b3c7854b3029315671cdebfe4f1
- last, but not least, an xtext-maven-plugin test for an Xbase language https://github.com/LorenzoBettini/xtext/commit/21b24acd317a99fdfcf4a5d09fa2518fbee27895 note that I had to force a more recent version of JDT supporting Java 21 in the plugin configuration in the IT project; without the updated JDT version the test fails as expected

The next step would be to update our dev BOM with the new JDT code. This will require Java 17 as the minimal version, unless we set the profile in the BOM switching the JDT version according to the current Java, as I had experimented here https://github.com/LorenzoBettini/maven-bom-profiles-example

**Comment by szarnekow** [comment]
> This will require Java 17 as the minimal version

If by any means possible,  I’d like to have Xtext 2.35 released with Java 11 bytecode. If that contradicts the ability to consume Java 21 bytecode, that would be very sad.

**Comment by LorenzoBettini** [comment]
@szarnekow there's no need to switch to Java 17 as requirement, and we can still ship Java 11, if we want to.

My experiments are meant to be sure we can consume Java 21 stuff (if we can consume records we should be safe in that respect).

Mostly, there are tests in my fork's branch.
The only real change in the code base is this one https://github.com/LorenzoBettini/xtext/commit/45519f52c316461b4156142f7095cfb0575c6379 but that method has been available for a while and should work also with our current JDT versions.

In Eclipse, as long as one uses the latest JDT there's no problem accessing Java records.

Concerning Maven, if we don't want to update the BOM, it is enough to document to add the more recent JDT as explicit dependencies https://github.com/LorenzoBettini/xtext/commit/21b24acd317a99fdfcf4a5d09fa2518fbee27895#diff-88e2b37f635d0790f54a2d314a7b4cb75652b846440f6d878a709a5a7910baadR55 if one want to access Java records.

To make Maven users easier, we could use the profile "trick" in the BOM I mentioned above. In my experiment, I switch to an older version of JDT when the current JDK is not Java 17 https://github.com/LorenzoBettini/maven-bom-profiles-example/blob/master/bom-example/pom.xml#L28. But of course we could do the opposite: switch to a newer version of jdt if the current JDK is at least Java 21 ;)

**Comment by cdietrich** [comment]
unfortunately i dont find the time to experiment with gradle and maven

**Comment by LorenzoBettini** [comment]
@szarnekow @cdietrich some updates:

- I applied the profile in the BOM, automatically activated when using Java 21, switching to the recent version of JDT core and it works: the xtext-maven-plugin test now works without explicit jdt dependencies: https://github.com/LorenzoBettini/xtext/commit/3e7d608bc802d0971cc6c6c67fe5df6ab67ad08c Thus, for previous versions of Java nothing changes and we can keep producing Java 11 bytecode
- I added an xtend-maven-plugin test that also works: https://github.com/LorenzoBettini/xtext/commit/315cbe86e74fce60117d356c1d413c09d1661142
- Finally, I added an Xtend validation test that correctly detects a record as a final class: https://github.com/LorenzoBettini/xtext/commit/6026fd427af100667f7a4a458803db283474f4a9

Concerning gradle, if it doesn't pick the Maven profile correctly, it's just a matter of forcing a more recent version of jdt.core in the xtend-maven-plugin configuration

**Comment by cdietrich** [comment]
the problem is gradle does not use the xtend-maven-plugin.
but its own.
there is also the xbase usecase.
dont know if and how this affects the xtext-maven-plugin

**Comment by LorenzoBettini** [comment]
> the problem is gradle does not use the xtend-maven-plugin. but its own.

then, the updated jdt core versions should be specified there

> there is also the xbase usecase. dont know if and how this affects the xtext-maven-plugin

please, see my comments above: I added tests in that respect, both for xtext-maven-plugin and xtend-maven-plugin

**Comment by cdietrich** [comment]
so you propose doing the override in the gradle wizard and write a "what to do in xbase case" guide for gradle

**Comment by LorenzoBettini** [comment]
> so you propose doing the override in the gradle wizard and write a "what to do in xbase case" guide for gradle

Yes, that would be a good strategy!

**Comment by cdietrich** [comment]
but the experience shows this is very very very very very error prone

**Comment by LorenzoBettini** [comment]
That's the best we can do. In the worst case, gradle users will not be able to use Java 21, which is the current state anyway. For the rest, nothing will be broken.

**Comment by cdietrich** [comment]
yeah but then we need an error / warning in the wizard etc...

thus my preference regarding sebastians proposal:

- keep java 11 for 2.35
- ditch it for 2.36

**Comment by LorenzoBettini** [comment]
Please see my comments above. My proposed solution would allow us to keep Java 11 while still providing support for Java 21 for those who want to move to Java 21.

**Comment by ArneDeutsch** [comment]
Java 21 is supported since 2.35.0. Can this be closed or what are the points still open that need to be done?

**Comment by cdietrich** [comment]
@LorenzoBettini afaik this needs your open pr for actual support with no workarounds in gradle

**Comment by LorenzoBettini** [comment]
> @LorenzoBettini afaik this needs your open pr for actual support with no workarounds in gradle

Ah, you mean the one about Java 17 right?

**Comment by cdietrich** [comment]
> Ah, you mean the one about Java 17 right?

yes
